File: codewar/elixir/007.exs
defmodule MorseCode do
  # use MorseCode.Constants # Defines @morse_codes
  @morse_codes %{}
  
  def get_dots(bits, acc \\ "") do
    case bits do
      "" -> acc
      "0" <> tail -> 
        if acc == "" do
          get_dots(tail)
        else
          get_dots("", acc)
        end
      "1" <> tail ->
        acc = acc <> "1"
        get_dots(tail, acc)
    end
  end

  def decode_bits(bits, morse_codes, result \\ "", acc \\ "", acc_bit \\ "") do
    # morse_codes |> IO.inspect
    case bits do
      "" -> result
      "0" <> tail -> 
        result = if acc_bit != "0" do
          result <> (morse_codes |> Map.get(acc, ""))
        else
          result
        end
        acc = if acc_bit != "0" do
          "0"
        else
          acc <> "0"
        end
        acc_bit = "0"
        decode_bits(tail, morse_codes, result, acc, acc_bit)
      "1" <> tail ->
        result = if acc_bit != "1" do
          result <> (morse_codes |> Map.get(acc, ""))
        else
          result
        end
        acc = if acc_bit != "1" do
          "1"
        else
          acc <> "1"
        end
        acc_bit = "1"
        decode_bits(tail, morse_codes, result, acc, acc_bit)
    end
  end

  def decode(code) do
    dots = with bits <- get_dots(code),
         bits_length <- bits |> String.length do
      if rem(bits_length, 3) == 0 do
        bits |> String.slice(0..div(bits_length, 3)-1) 
      else
        bits
      end
    end
    dashs = dots |> String.duplicate(3)
    dots_length = dots |> String.length
    morse_codes = Map.merge(
      @morse_codes,
      %{
        String.duplicate("0", dots_length) => "",
        String.duplicate("0", dots_length*3) => " ",
        String.duplicate("0", dots_length*7) => "  ",
        dashs => "-",
        dots => ".",
      }
    )
    morse_codes |> IO.inspect
    code
    |> String.trim("0")
    |> decode_bits(morse_codes)
    |> String.split("  ")
    |> Enum.map(
      fn token -> do
        token
        |> String.split()
        |> Enum.map(
          fn word -> morse_codes |> Map.get(word, "") end
        )
        |> Enum.join
      end
    )
    |> Enum.join(" ")
  end
end


"1100110011001100000011000000111111001100111111001111110000000000000011001111110011111100111111000000110011001111110000001111110011001100000011" |> MorseCode.decode |> IO.inspect

File: codewar/elixir/007.py
MORSE_CODE = {'.-': 'A', '-...': 'B', '-.-.': 'C', '-..': 'D', '.': 'E', '..-.': 'F', '--.': 'G', '....': 'H', '..': 'I', '.---': 'J', '-.-': 'K', '.-..': 'L', '--': 'M', '-.': 'N', '---': 'O', '.--.': 'P', '--.-': 'Q', '.-.': 'R', '...': 'S', '-': 'T', '..-': 'U', '...-': 'V', '.--': 'W', '-..-': 'X', '-.--': 'Y', '--..': 'Z', '-----': '0', '.----': '1', '..---': '2', '...--': '3', '....-': '4', '.....': '5', '-....': '6', '--...': '7', '---..': '8', '----.': '9', '.-.-.-': '.', '--..--': ',', '..--..': '?', '.----.': "'", '-.-.--': '!', '-..-.': '/', '-.--.': '(', '-.--.-': ')', '.-...': '&', '---...': ':', '-.-.-.': ';', '-...-': '=', '.-.-.': '+', '-....-': '-', '..--.-': '_', '.-..-.': '"', '...-..-': '$', '.--.-.': '@', '...---...': 'SOS'}

def get_replace_map_0(bits):
    acc = ""
    for bit in bits:
        if bit == "1":
            acc += bit
        if bit == "0" and acc != "":
            break
    acc_length = len(acc)
    dots = acc[0:acc_length//3] if acc_length % 3 == 0 else acc
    dashs = dots * 3
    return {
        dashs: "-",
        dots: ".",
        "0"*len(dots)*7: "  ",
        "0"*len(dots)*3: " ",
        "0"*len(dots): "",
    }

def get_replace_map(bits):
    min_one = min(map(lambda token: len(token), bits.replace('0',' ').split()))
    min_zero = min(map(lambda token: len(token), bits.replace('1',' ').split()))
    dots = (min(min_one, min_zero) if min_one * min_zero != 0 else min_one+min_zero) * '1'
    print(dots)
    dashs = dots * 3
    return {
        dashs: "-",
        dots: ".",
        "0"*len(dots)*7: "  ",
        "0"*len(dots)*3: " ",
        "0"*len(dots): "",
    }

def decode_bits(bits):
    replace_map = get_replace_map(bits)
    for k in replace_map:
        bits = bits.replace(k, replace_map[k])
    print('bits')
    print(bits)
    return bits

def decode_morse(morseCode):
    # ToDo: Accept dots, dashes and spaces, return human-readable message
    # morseCode = decode_bits(morseCode)
    codes = morseCode.split()
    for code in sorted(codes, key=lambda v: len(v), reverse=True):
        morseCode = morseCode.replace(code, MORSE_CODE[code])
    return ' '.join(
      map(lambda string: string.replace(' ', ''), morseCode.split('  '))
    )

# print(decode_bits("1100110011001100000011000000111111001100111111001111110000000000000011001111110011111100111111000000110011001111110000001111110011001100000011"))
print(decode_morse(decode_bits("1100110011001100000011000000111111001100111111001111110000000000000011001111110011111100111111000000110011001111110000001111110011001100000011")))
print(decode_morse(decode_bits("111000111")))
print(decode_morse(decode_bits('111000000000111')))
